ml-proto/sexp.py

```diff
@@ -0,0 +1,234 @@
+#!/usr/bin/env python
+
+class SExp(object):
+    def __init__(self, thing, begin=-1, end=-1):
+        assert (begin == -1) == (end == -1)
+        assert not isinstance(thing, SExp)
+        self.thing = thing
+        self.begin = begin
+        self.end = end
+
+    def __repr__(self):
+        return str(self.thing) + ':' + str(self.begin) + ':' + str(self.end)
+
+    def __str__(self):
+        return str(self.thing)
+
+    def has_location(self):
+        return self.begin != -1
+
+    def is_operator(self, name):
+        return isinstance(self.thing, list) and len(self.thing) > 0 and self.thing[0].thing == name
+
+    def num_operands(self):
+        return len(self.thing) - 1
+
+    def get_operand(self, index):
+        return self.thing[index + 1]
+
+    def strip(self):
+        if isinstance(self.thing, list):
+            for item in self.thing:
+                item.strip()
+        self.begin = -1
+        self.end = -1
+
+def isspace(c):
+    return c == ' ' or c == '\t' or c == '\n'
+
+def isdigit(c):
+    return c >= '0' and c <= '9'
+
+def isalnum(c):
+    return isdigit(c) or (c >= 'a' and c <= 'z') or (c >= 'A' and c <= 'Z')
+
+def skip_whitespace_and_comments(s, i):
+    l = len(s)
+    while i < l:
+        if isspace(s[i]):
+            i += 1
+            # fixme
+            #while i < l and isspace(s[i]):
+                #i += 1
+        elif s[i] == ';' and i+1 < l and s[i+1] == ';':
```

Comment:
What about landing this in a separate repo? The concern here is to avoid
specifying a particular testing helper language as part of the official
definition of WebAssembly.

On Wed, Nov 4, 2015 at 12:15 PM, Patrick Uftring <notifications@github.com>
wrote:

> In ml-proto/sexp.py
> <https://github.com/WebAssembly/spec/pull/155#discussion_r43933169>:
>
> > +
> > +def isdigit(c):
> > +    return c >= '0' and c <= '9'
> > +
> > +def isalnum(c):
> > +    return isdigit(c) or (c >= 'a' and c <= 'z') or (c >= 'A' and c <= 'Z')
> > +
> > +def skip_whitespace_and_comments(s, i):
> > +    l = len(s)
> > +    while i < l:
> > +        if isspace(s[i]):
> > +            i += 1
> > +            # fixme
> > +            #while i < l and isspace(s[i]):
> > +                #i += 1
> > +        elif s[i] == ';' and i+1 < l and s[i+1] == ';':
>
> I don't know if anybody uses them, but
>
> (* This type of comment *)
>
> Isn't handled by this code. They seem to work, but if I don't close a
> bracket:
>
> (* (Like this *)
>
> It raises an index error.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/WebAssembly/spec/pull/155/files#r43933169>.
>